src/lib/core/error/error-options.ts

```diff
@@ -7,27 +7,24 @@
  */
 
 import {InjectionToken} from '@angular/core';
-import {FormControl, FormGroupDirective, NgForm} from '@angular/forms';
+import {FormGroupDirective, NgForm, NgControl} from '@angular/forms';
 
 /** Injection token that can be used to specify the global error options. */
 export const MD_ERROR_GLOBAL_OPTIONS = new InjectionToken<ErrorOptions>('md-error-global-options');
 
 export type ErrorStateMatcher =
```

Comment:
So I was thinking about maybe refactoring some of this error state matcher stuff because @g1shin recently had to implement a context-sensitive one for the stepper (#6117) and it felt a bit messier than it needed to be... What do you think about something like this:

```ts
export interface ErrorStateMatcher {
  isErrorState(...): boolean;
} 

export class DefaultErrorStateMatcher {
  isErrorState(...) { ... }
}

export class ShowOnDirtyErrorStateMatcher {
  isErrorState(...) { ... }
}
```

And then in error module:
```ts
@NgModule({
  providers: [{provide: ErrorStateMatcher, useClass: DefaultErrorStateMatcher}]
})
```

And finally in input/select/etc:
```ts
constructor(errorStateMatcher: ErrorStateMatcher) {
  console.log(errorStateMatcher.isErrorState(...));
}
```

I like this because:
1) We don't have to worry about the function being ripped off the class and losing its context
2) We don't have to worry about falling back to the defaultErrorStateMatcher in every component since we provide that in the error module
3) The code to inject in the constructor doesn't need a whole bunch of `@Optional() @Inject(...)` stuff